### PR TITLE
s2455 - retain viewer column state from the previous search

### PR DESCRIPTION
**Pull request by jburke-cadc**

### CODE REVIEW

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js [hunk 1]
```diff
@@ -1575,6 +1627,24 @@
       }
     }
 
+    /**
```
Comment (at88mph) [inline caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: Please just delete this if it's not used.

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js [hunk 2]
```diff
@@ -1534,37 +1552,71 @@
...
-        .getDefaultUnitTypes()
 
-      for (var columnName in unitTypes) {
+      //
+      var allColumnOptions = this._getActiveForm().getConfiguration().getColumnOptions()
+      var allUnitTypes = {}
+      for (var columnOption in allColumnOptions) {
+        var column = allColumnOptions[columnOption]
+        if (column.header)
+        {
```
Comment (at88mph) [inline caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: Sorry to be picky.  For consistency, please put if/else braces inline.

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js [hunk 3]
```diff
@@ -1080,29 +1089,38 @@
...
+        resultsVOTV.setDisplayColumns(prevDisplayedColumns)
+        resultsVOTV.setUpdatedColumnSelects(prevColumnSelects)
 
-        // Set the default columns.
+        // Set the default columns and units.
         this._setDefaultColumns(resultsVOTV)
         this._setDefaultUnitTypes(resultsVOTV)
 
+        // Set default sort column and direction.
+        if (prevSortOptions['sortcol']) {
```
Comment (at88mph) [inline caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: This the same as `if (prevSortOptions.hasOwnProperty('sortcol')) {`, which is faster.

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js (hunk 1 above)
Comment (jburke-cadc) [reply caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: Deleted

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js (hunk 2 above)
Comment (jburke-cadc) [reply caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: Fixed

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js (hunk 3 above)
Comment (jburke-cadc) [reply caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: Updated.

File: caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js [hunk 4]
```diff
@@ -804,20 +805,27 @@
        * Form validation succeeded.
        */
       var onFormValid = function (eventData, args) {
+        var preserveColumnState = false
         var prevColumns = []
         var prevDisplayedColumns = []
         var prevColumnSelects = {}
         var prevSortOptions = {}
+        var currentCollections = $('#Observation\\.collection').val()
```
Comment (at88mph) [inline caom2-search-server/src/main/resources/META-INF/resources/js/cadc.search.app.js]: This only handles the CAOM-2 form.  See the `form.getConfiguration().getDownloadAccessKey()` as an example of a way of making the field name generic and configured per Form Configuration.

**Comment by at88mph** [comment]
The latest Firefox has an issue when clicking fields.  Merging this for now, but will address the Firefox issue immediately.

**Pull request merged**